### ISSUE TITLE
Backstop is running but not creating enough screenshots

### ISSUE THREAD

**Issue by VolleMelk** [garris/BackstopJS#581]
So have have this basic backstop.json:

```json
{
  "id": "project",
  "viewports": [
    {
      "label": "phone",
      "width": 320,
      "height": 480
    },
    {
      "label": "tablet",
      "width": 1024,
      "height": 768
    }
  ],
  "onBeforeScript": "chromy/onBefore.js",
  "onReadyScript": "chromy/onReady.js",
  "scenarios": [],
  "paths": {
    "bitmaps_reference": "backstop_data/bitmaps_reference",
    "bitmaps_test": "backstop_data/bitmaps_test",
    "engine_scripts": "backstop_data/engine_scripts",
    "html_report": "backstop_data/html_report",
    "ci_report": "backstop_data/ci_report"
  },
  "report": [],
  "engine": "chrome",
  "engineFlags": [],
  "asyncCaptureLimit": 1,
  "asyncCompareLimit": 50,
  "debug": true,
  "debugWindow": true
}
```

`scenarios` is empty. I'm requiring my configuration inside my gulpfile, and dynamically creating my scenarios:

```javascript
const fs = require('fs');
const backstopJs = require('backstopjs');
const backstopJsConfig = require('./backstop.json');

gulp.task('backstop_reference', function () {
  return backstop('reference');
});
gulp.task('backstop_test', function () {
  return backstop('test');
});

function backstop(action) {
  const dir = 'build/library/components/preview';

  var files = fs.readdirSync(dir)
    .filter(function (file) {
      return fs.statSync(path.join(dir, file)).isFile();
    });

  console.log(`Files: ${files.join('\n')}`);

  return backstopJs(action, Object.assign(backstopJsConfig, {
    scenarios: files.map(f => ({ label: f, url: path.join(dir, f) }))
  }))
  .slice(0, 10);
}
```

This works like a charm. The debug console.log is confirming that I have files to be scanned. During my troubleshooting I'm only using the first 10 files. BackstopJS is only generating 3 files:

* project_badgehtml_0_document_0_phone.png
* project_button-grouphtml_0_document_1_tablet
* project_buttonhtml_0_document_0_phone

Console output:

```
D:\git\project [backstopjs +3 ~4 -251 !]> gulp backstop_reference
[15:07:56] Using gulpfile D:\git\project\gulpfile.js
[15:07:56] Starting 'backstop_reference'...
[15:07:56] Files: aanvragen.html
accordion.html
alert.html
anchor.html
badge.html
breadcrumbs.html
button-group.html
button.html
carousel.html
checken.html
Loading config:  D:\git\project\backstop.json

COMMAND | Executing core for `reference`
  clean | backstop_data/bitmaps_reference was cleaned.
createBitmaps | Selected 10 of 10 scenarios.
Starting Chromy: port:9222 --disable-gpu,--force-device-scale-factor=1,--disable-infobars=true,--window-size=320,480
CREATING NEW REFERENCE FILES
Cookie state restored with: []
SCENARIO > aanvragen.html
Starting Chromy: port:9223 --disable-gpu,--force-device-scale-factor=1,--disable-infobars=true,--window-size=1024,768
CREATING NEW REFERENCE FILES
Cookie state restored with: []
SCENARIO > aanvragen.html
Starting Chromy: port:9224 --disable-gpu,--force-device-scale-factor=1,--disable-infobars=true,--window-size=320,480
CREATING NEW REFERENCE FILES
Cookie state restored with: []
SCENARIO > accordion.html
Starting Chromy: port:9225 --disable-gpu,--force-device-scale-factor=1,--disable-infobars=true,--window-size=1024,768
CREATING NEW REFERENCE FILES
Cookie state restored with: []
SCENARIO > accordion.html
Starting Chromy: port:9226 --disable-gpu,--force-device-scale-factor=1,--disable-infobars=true,--window-size=320,480
CREATING NEW REFERENCE FILES
Cookie state restored with: []
SCENARIO > alert.html
Starting Chromy: port:9227 --disable-gpu,--force-device-scale-factor=1,--disable-infobars=true,--window-size=1024,768
CREATING NEW REFERENCE FILES
Cookie state restored with: []
SCENARIO > alert.html
Starting Chromy: port:9228 --disable-gpu,--force-device-scale-factor=1,--disable-infobars=true,--window-size=320,480
CREATING NEW REFERENCE FILES
Cookie state restored with: []
SCENARIO > anchor.html
Starting Chromy: port:9229 --disable-gpu,--force-device-scale-factor=1,--disable-infobars=true,--window-size=1024,768
CREATING NEW REFERENCE FILES
Cookie state restored with: []
SCENARIO > anchor.html
Starting Chromy: port:9230 --disable-gpu,--force-device-scale-factor=1,--disable-infobars=true,--window-size=320,480
CREATING NEW REFERENCE FILES
Cookie state restored with: []
SCENARIO > badge.html
9230 Chrome v64 detected.
9230 INFO >  BackstopTools have been installed.
9230HEAD >  <head>
    <meta charset="utf-8">

    <title>Default | project Component Library - Component preview</title>

    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">

    <link rel="stylesheet" href="../../styles/project.css">
  </head>
9230BODY >  <body>
    <div class="container">
      <span class="badge">3</span>

    </div>

    <script src="https://code.jquery.com/jquery-3.2.1.js" integrity="sha256-DZAnKJ/6XZ9si04Hgrsxu/8s717jcIzLy3oi35EouyE=" crossorigin="anonymous"></script>
    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

    <script src="../../scripts/project.js"></script>


</body>
Starting Chromy: port:9231 --disable-gpu,--force-device-scale-factor=1,--disable-infobars=true,--window-size=1024,768
CREATING NEW REFERENCE FILES
Cookie state restored with: []
SCENARIO > badge.html
Starting Chromy: port:9232 --disable-gpu,--force-device-scale-factor=1,--disable-infobars=true,--window-size=320,480
CREATING NEW REFERENCE FILES
Cookie state restored with: []
SCENARIO > breadcrumbs.html
Starting Chromy: port:9233 --disable-gpu,--force-device-scale-factor=1,--disable-infobars=true,--window-size=1024,768
CREATING NEW REFERENCE FILES
Cookie state restored with: []
SCENARIO > breadcrumbs.html
Starting Chromy: port:9234 --disable-gpu,--force-device-scale-factor=1,--disable-infobars=true,--window-size=320,480
CREATING NEW REFERENCE FILES
Cookie state restored with: []
SCENARIO > button-group.html
(node:9020) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 SIGINT listeners added. Use emitter.setMaxListeners() to increase limit
Starting Chromy: port:9235 --disable-gpu,--force-device-scale-factor=1,--disable-infobars=true,--window-size=1024,768
CREATING NEW REFERENCE FILES
Cookie state restored with: []
SCENARIO > button-group.html
9235 Chrome v64 detected.
9235 INFO >  BackstopTools have been installed.
9235HEAD >  <head>
    <meta charset="utf-8">

    <title>Button Group | project Component Library - Component preview</title>

    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">

    <link rel="stylesheet" href="../../styles/project.css">
  </head>
9235BODY >  <body>
    <div class="container">
      <!-- Start: @button-group--m -->
<div class="btn-group " role="group" aria-label="Medium button group">
    <button type="button" class="btn btn-primary">Links</button>
    <button type="button" class="btn btn-primary">Midden</button>
    <button type="button" class="btn btn-primary">Rechts</button>
</div>
<!-- End: @button-group--m --><hr>

<!-- Start: @button-group--xs -->
<div class="btn-group btn-group-xs" role="group" aria-label="Extra small button group">
    <button type="button" class="btn btn-primary">Links</button>
    <button type="button" class="btn btn-primary">Midden</button>
    <button type="button" class="btn btn-primary">Rechts</button>
</div>
<!-- End: @button-group--xs --><hr>

<!-- Start: @button-group--s -->
<div class="btn-group btn-group-sm" role="group" aria-label="Small button group">
    <button type="button" class="btn btn-primary">Links</button>
    <button type="button" class="btn btn-primary">Midden</button>
    <button type="button" class="btn btn-primary">Rechts</button>
</div>
<!-- End: @button-group--s --><hr>

<!-- Start: @button-group--l -->
<div class="btn-group btn-group-lg" role="group" aria-label="Large button group">
    <button type="button" class="btn btn-primary">Links</button>
    <button type="button" class="btn btn-primary">Midden</button>
    <button type="button" class="btn btn-primary">Rechts</button>
</div>
<!-- End: @button-group--l --><hr>

    </div>

    <script src="https://code.jquery.com/jquery-3.2.1.js" integrity="sha256-DZAnKJ/6XZ9si04Hgrsxu/8s717jcIzLy3oi35EouyE=" crossorigin="anonymous"></script>
    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

    <script src="../../scripts/project.js"></script>


</body>
Starting Chromy: port:9236 --disable-gpu,--force-device-scale-factor=1,--disable-infobars=true,--window-size=320,480
CREATING NEW REFERENCE FILES
Cookie state restored with: []
SCENARIO > button.html
9236 Chrome v64 detected.
9236 INFO >  BackstopTools have been installed.
9236HEAD >  <head>
    <meta charset="utf-8">

    <title>Button | project Component Library - Component preview</title>

    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">

    <link rel="stylesheet" href="../../styles/project.css">
  </head>
9236BODY >  <body>
    <div class="container">
      <!-- Start: @button--default -->
<button type="button" class="btn btn-default">
    Versturen
</button>
<!-- End: @button--default --><hr>

<!-- Start: @button--primary -->
<button type="submit" class="btn btn-primary">
    Versturen
</button>
<!-- End: @button--primary --><hr>

<!-- Start: @button--disabled -->
<button type="submit" class="btn btn-default" disabled="">
    Versturen
</button>
<!-- End: @button--disabled --><hr>

<!-- Start: @button--with-icon -->
<button type="button" class="btn btn-default">
    <span class="fa fa-lock" aria-hidden="true"></span>
    <span class="sr-only">Vergrendelen</span>
</button>
<!-- End: @button--with-icon --><hr>

<!-- Start: @button--with-badge -->
<button type="button" class="btn btn-default">
    Versturen
    <span class="badge">3</span>
</button>
<!-- End: @button--with-badge --><hr>

<!-- Start: @button--with-icon-badge -->
<button type="button" class="btn btn-default">
    <span class="fa fa-envelope-o" aria-hidden="true"></span>
    Berichten
    <span class="badge">3</span>
</button>
<!-- End: @button--with-icon-badge --><hr>

    </div>

    <script src="https://code.jquery.com/jquery-3.2.1.js" integrity="sha256-DZAnKJ/6XZ9si04Hgrsxu/8s717jcIzLy3oi35EouyE=" crossorigin="anonymous"></script>
    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

    <script src="../../scripts/project.js"></script>


</body>
Starting Chromy: port:9237 --disable-gpu,--force-device-scale-factor=1,--disable-infobars=true,--window-size=1024,768
CREATING NEW REFERENCE FILES
Cookie state restored with: []
SCENARIO > button.html
Starting Chromy: port:9238 --disable-gpu,--force-device-scale-factor=1,--disable-infobars=true,--window-size=320,480
CREATING NEW REFERENCE FILES
Cookie state restored with: []
SCENARIO > carousel.html
Starting Chromy: port:9239 --disable-gpu,--force-device-scale-factor=1,--disable-infobars=true,--window-size=1024,768
CREATING NEW REFERENCE FILES
Cookie state restored with: []
SCENARIO > carousel.html
Starting Chromy: port:9240 --disable-gpu,--force-device-scale-factor=1,--disable-infobars=true,--window-size=320,480
CREATING NEW REFERENCE FILES
Cookie state restored with: []
SCENARIO > checken.html
Starting Chromy: port:9241 --disable-gpu,--force-device-scale-factor=1,--disable-infobars=true,--window-size=1024,768
CREATING NEW REFERENCE FILES
Cookie state restored with: []
SCENARIO > checken.html

Run `$ backstop test` to generate diff report.

      COMMAND | Command `reference` sucessfully executed in [13.939s]
[15:08:10] Finished 'backstop_reference' after 14 s
D:\git\project [backstopjs +3 ~4 -251 !]>
```

I haven't got the faintest what to do.

I'm running Windows 10, Node 8.8.1, npm 5.5.1, backstopjs 3.0.31 and Chrome 62

**Comment by VladimirPittner** [comment]
Hi,
can you provide the files with scenarios?
Are you sure each scenario have unique label?

/V

**Comment by VolleMelk** [comment]
My generated configuration right before it's passed to backstopjs:

```
{ id: 'dso',
  viewports:
   [ { label: 'phone', width: 320, height: 480 },
     { label: 'tablet', width: 1024, height: 768 } ],
  onBeforeScript: 'chromy/onBefore.js',
  onReadyScript: 'chromy/onReady.js',
  scenarios:
   [ { label: 'aanvragen.html',
       url: 'build\\library\\components\\preview\\aanvragen.html' },
     { label: 'accordion.html',
       url: 'build\\library\\components\\preview\\accordion.html' },
     { label: 'alert.html',
       url: 'build\\library\\components\\preview\\alert.html' },
     { label: 'anchor.html',
       url: 'build\\library\\components\\preview\\anchor.html' },
     { label: 'badge.html',
       url: 'build\\library\\components\\preview\\badge.html' },
     { label: 'breadcrumbs.html',
       url: 'build\\library\\components\\preview\\breadcrumbs.html' },
     { label: 'button-group.html',
       url: 'build\\library\\components\\preview\\button-group.html' },
     { label: 'button.html',
       url: 'build\\library\\components\\preview\\button.html' },
     { label: 'carousel.html',
       url: 'build\\library\\components\\preview\\carousel.html' },
     { label: 'checken.html',
       url: 'build\\library\\components\\preview\\checken.html' },
     { label: 'dropdown-button.html',
       url: 'build\\library\\components\\preview\\dropdown-button.html' },
     { label: 'dso-select.html',
       url: 'build\\library\\components\\preview\\dso-select.html' },
     { label: 'footer.html',
       url: 'build\\library\\components\\preview\\footer.html' },
     { label: 'form-horizontal.html',
       url: 'build\\library\\components\\preview\\form-horizontal.html' },
     { label: 'form-inline.html',
       url: 'build\\library\\components\\preview\\form-inline.html' },
     { label: 'form-vertical.html',
       url: 'build\\library\\components\\preview\\form-vertical.html' },
     { label: 'grid.html',
       url: 'build\\library\\components\\preview\\grid.html' },
     { label: 'header.html',
       url: 'build\\library\\components\\preview\\header.html' },
     { label: 'help-block.html',
       url: 'build\\library\\components\\preview\\help-block.html' },
     { label: 'highlight-box.html',
       url: 'build\\library\\components\\preview\\highlight-box.html' },
     { label: 'horizontal-group-checkboxes.html',
       url: 'build\\library\\components\\preview\\horizontal-group-checkboxes.html' },
     { label: 'horizontal-group-confirm.html',
       url: 'build\\library\\components\\preview\\horizontal-group-confirm.html' },
     { label: 'horizontal-group-input.html',
       url: 'build\\library\\components\\preview\\horizontal-group-input.html' },
     { label: 'horizontal-group-radios.html',
       url: 'build\\library\\components\\preview\\horizontal-group-radios.html' },
     { label: 'horizontal-group-select.html',
       url: 'build\\library\\components\\preview\\horizontal-group-select.html' },
     { label: 'horizontal-group-static.html',
       url: 'build\\library\\components\\preview\\horizontal-group-static.html' },
     { label: 'horizontal-group-textarea.html',
       url: 'build\\library\\components\\preview\\horizontal-group-textarea.html' },
     { label: 'icon.html',
       url: 'build\\library\\components\\preview\\icon.html' },
     { label: 'image.html',
       url: 'build\\library\\components\\preview\\image.html' },
     { label: 'inline-group-input.html',
       url: 'build\\library\\components\\preview\\inline-group-input.html' },
     { label: 'inline-group-static.html',
       url: 'build\\library\\components\\preview\\inline-group-static.html' },
     { label: 'input-group.html',
       url: 'build\\library\\components\\preview\\input-group.html' },
     { label: 'jumbotron.html',
       url: 'build\\library\\components\\preview\\jumbotron.html' },
     { label: 'label.html',
       url: 'build\\library\\components\\preview\\label.html' },
     { label: 'lists.html',
       url: 'build\\library\\components\\preview\\lists.html' },
     { label: 'logo.html',
       url: 'build\\library\\components\\preview\\logo.html' },
     { label: 'menu.html',
       url: 'build\\library\\components\\preview\\menu.html' },
     { label: 'navs-item.html',
       url: 'build\\library\\components\\preview\\navs-item.html' },
     { label: 'navs.html',
       url: 'build\\library\\components\\preview\\navs.html' },
     { label: 'pager.html',
       url: 'build\\library\\components\\preview\\pager.html' },
     { label: 'pagination.html',
       url: 'build\\library\\components\\preview\\pagination.html' },
     { label: 'panel.html',
       url: 'build\\library\\components\\preview\\panel.html' },
     { label: 'paragraph.html',
       url: 'build\\library\\components\\preview\\paragraph.html' },
     { label: 'progress-bar.html',
       url: 'build\\library\\components\\preview\\progress-bar.html' },
     { label: 'progress-block.html',
       url: 'build\\library\\components\\preview\\progress-block.html' },
     { label: 'split-button-dropdown.html',
       url: 'build\\library\\components\\preview\\split-button-dropdown.html' },
     { label: 'streamer.html',
       url: 'build\\library\\components\\preview\\streamer.html' },
     { label: 'table.html',
       url: 'build\\library\\components\\preview\\table.html' },
     { label: 'vertical-group-checkboxes.html',
       url: 'build\\library\\components\\preview\\vertical-group-checkboxes.html' },
     { label: 'vertical-group-confirm.html',
       url: 'build\\library\\components\\preview\\vertical-group-confirm.html' },
     { label: 'vertical-group-input.html',
       url: 'build\\library\\components\\preview\\vertical-group-input.html' },
     { label: 'vertical-group-radios.html',
       url: 'build\\library\\components\\preview\\vertical-group-radios.html' },
     { label: 'vertical-group-select.html',
       url: 'build\\library\\components\\preview\\vertical-group-select.html' },
     { label: 'vertical-group-static.html',
       url: 'build\\library\\components\\preview\\vertical-group-static.html' },
     { label: 'vertical-group-textarea.html',
       url: 'build\\library\\components\\preview\\vertical-group-textarea.html' },
     { label: 'wizard.html',
       url: 'build\\library\\components\\preview\\wizard.html' } ],
  paths:
   { bitmaps_reference: 'backstop_data/bitmaps_reference',
     bitmaps_test: 'backstop_data/bitmaps_test',
     engine_scripts: 'backstop_data/engine_scripts',
     html_report: 'backstop_data/html_report',
     ci_report: 'backstop_data/ci_report' },
  report: [],
  engine: 'chrome',
  engineFlags: [],
  asyncCaptureLimit: 1,
  asyncCompareLimit: 50,
  debug: true,
  debugWindow: true }
```

**Comment by lancetharper** [comment]
I have a similar problem in both reference and test file generation.

I have 4 breakpoints specified --> phone, tablet_v, tablet_h, desktop. So every selector should get 4 screenshots.

2 of my selectors only get 3 screenshots. One is missing "phone" the other is missing "tablet_h". 

They are missing in both reference and test, but it knows that they should be compared because it fails when it tries to do the diff.

This just started failing after I updated from 3.0.25

**Comment by garris** [comment]
@lancetharper  can you post a gist config and session log?   Also,  what version of everything are you running.   BackstopJS should be 3.0.31 or greater.

**Comment by lancetharper** [comment]
Node: 8.8.1
npm: 5.5.1
chrome: 62
backstop: 3.0.31

I updated my node/npm in case that was the issue and now it's even more inconsistent. Sometimes it's missing 6 or 7 screenshots now. 

config (generated) and output gist [here](https://gist.github.com/lancetharper/7f823789f02a8a57740a92a52c1f5ac4)

**Comment by lancetharper** [comment]
@garris After I had a clean run, I made a change just to double check that it was working. It failed where it should have but also 1 place it shouldn't have (didn't have a test screenshot but it did have a reference).

I checked activity monitor to see if there were any chrome processes that weren't exited -- there was one. I killed it and ran again and everything worked as expected.

So it seems the issue I'm seeing is chromy/backstop not killing chrome processes sometimes...?

**Comment by garris** [comment]
@lancetharper Glad you were able to get it working!  Yes, it's totally possible a zombie chrome process can be the issue.  Most unhandled error states that could lead to this issue were fixed in 3.0.27... 
 https://github.com/garris/BackstopJS/commit/8e0074f8fd6366f87d4743099735e79ddcc40b1f#diff-9e67e5c53c6a24f5b7f26f857fff03eaR174

After that fix I haven't seen the issue again -- even when killing the process manually.   That said, you could probably still get a zombie if a bad enough error occurs (like if chrome itself freezes).   To double make sure this doesn't happen at work, I've added a chrome-headless forced kill command in front of the npm script which calls various backstop tests e.g. `pkill -f '(chrome)?(--headless)'; backstop test`

Thanks for posting!

**Comment by garris** [comment]
published to npm 3.0.32

**Comment by VolleMelk** [comment]
It appears the new version resolved my problems as well, thanks @garris !
